src/main/java/edu/ucsd/workflow/RemoveRedundantTSVRows.java

```diff
@@ -0,0 +1,220 @@
+	 * Constants
+	 *========================================================================*/
+	private static final String USAGE = "java -jar CCMSWorkflowUtils.jar" +
+		"\n\t-input <InputFile> (TSV file with header)" +
+		"\n\t-output <OutputFile> (Input file with redundant rows removed)" +
+		"\n\t-columns <ColumnHeaders> (Comma-separated list of columns " +
+		"for which only the first row per unique combination of these " +
+		"columns should be retained)";
+
+	/*========================================================================
+	 * Public interface methods
+	 *========================================================================*/
+	public static void main(String[] args) {
+		RemoveRedundantTSVRowsOperation removeRedundantTSVRows = extractArguments(args);
+		if (removeRedundantTSVRows == null)
+			die(USAGE);
+		// remove redundant rows in the TSV
+		BufferedReader input = null;
+		PrintWriter output = null;
+		Throwable thrownError = null;
+		int linesInInputFile = 1;
+		int linesInOutputFile = 1;
+		try {
+			input = new BufferedReader(new FileReader(
+				removeRedundantTSVRows.inputFile));
+			// set up output file writer
+			output = new PrintWriter(removeRedundantTSVRows.outputFile);
+			System.out.println(String.format(
+				"Writing the non-redundant rows in file \"%s\" to output file \"%s\".",
+				removeRedundantTSVRows.inputFile,
+				removeRedundantTSVRows.outputFile));
+			// read header
+			String headerStr = input.readLine();
+			String[] header = headerStr.trim().split("\t");
+			String columnsStr = "";
+			Set<List<String>> uniqueLineIDs = new HashSet<>();
+			for (int i=0; i<header.length; i++) {
+				header[i] = header[i].trim();
+			}
+			for (int i=0; i<removeRedundantTSVRows.columns.length; i++) {
+				String column = removeRedundantTSVRows.columns[i];
+				if (!(Arrays.asList(header).contains(column))) {
+					throw new IllegalArgumentException(
+						String.format("Column \"%s\" was not found in input header.",
+						column));
+				}
+				columnsStr += column + ", ";
+			}
+			if (columnsStr.endsWith(", "))
+				columnsStr = columnsStr.substring(0, columnsStr.length() - ", ".length());
+			System.out.println(String.format(
+				"Retaining the first row in input file per unique combination of " +
+				"the following columns: %s", columnsStr));
+			output.println(headerStr);
+			// read the remaining lines, and write the non-redundant ones to the output file
+			String lineStr = null;
+			while ((lineStr = input.readLine()) != null) {
+				linesInInputFile++;
+				String[] line = lineStr.trim().split("\t");
+				List<String> uniqueLineID = new ArrayList<>();
+				for (int i=0; i<line.length; i++) {
+					if (Arrays.asList(removeRedundantTSVRows.columns).contains(
+						header[i])) {
+						uniqueLineID.add(line[i]);
+					}
+				}
+				if (!(uniqueLineIDs.contains(uniqueLineID))) {
+					linesInOutputFile++;
+					output.println(lineStr);
+					uniqueLineIDs.add(uniqueLineID);
+				}
```

Comment:
The logic here is good, but I don't quite understand the point of specifying a unique set of columns on the command line, and validating uniqueness against that. I guess the idea is that you're not just throwing out redundant lines, but really only validating against a particular set of columns? So for example, even if each row has 100 columns, you might want to filter out columns having duplicates of only columns A. B, and C? Similar to a SQL "group by" query. Because simply comparing line uniqueness with a set of strings is obviously much more efficient.

Reply:
The idea is that, for example, there is a PSM file that has the exact same line twice, but with a different protein in the second line. If we don't care about the protein, we can reduce the size of the file by omitting the second line. The number and name of columns needs to be variable because, say, for one tool, we want one row per cluster index, whereas for another tool, we want one row per combination of filename and scan.

I thought about using concatenated strings instead of a list of strings to store the important columns in each line, but my issue with that was that could be weird edge cases such as
column A="AB", column B="CD" is equivalent to column A="A", column B="BCD".
The same problem can happen even with delimiters between the strings, unless there's some delimiter that will never occur in an input file.